### ISSUE TITLE
[META] Automatic Updates

### ISSUE THREAD

**Issue by klonos** [backdrop/backdrop-issues#2018]
Placeholder issue for any tasks related to Backdrop CMS for https://github.com/mbaynton/cms-autoupdate-design with task list to be eventually formed for any actionable things specific to us.

Everyone's opinion is welcomed, but we specifically seek opinions from the PMC members.

**Comment by laryn** [comment]
I thought there was further discussion on the topic that I don't see here, but I liked where the discussion was going in terms of:

 - A setting to allow folks to turn it on or off.
 - A separate security patch branch that only fixed the security issue without forcing other upgrades, including for previous versions (e.g. a security patch for 1.8.x that wouldn't force an upgrade to 1.9.x) to minimize potential for breaking updates through this system.

**Comment by olafgrabienski** [comment]
Reread the article on [Automatic Update Technical Design Planning](https://github.com/mbaynton/cms-autoupdate-design), which seems still worth thinking about. Additionally to the items pointed out by @laryn, I also like the concept of determination which of the offered updates a site receives automatically based on security risk scores.

**Comment by olafgrabienski** [comment]
Now as we've just experienced a series of critical Backdrop security releases, let's revive the idea of automatic updates. I guess it's not easy to plan and implement such a feature but it could be a huge improvement which would match the philosophy of Backdrop perfectly. Here some goals of automatic (security) updates from my perspective:

- Keep more Backdrop sites secure and alive.
- Save time, so that not only agencies but also individuals (hobbyists, voluntaries, freelancers) are able to maintain Backdrop sites, or to maintain more Backdrop sites. With "save time" I mean
  - the span of time which I need before a site gets eventually exploited, and
  - the working hours which I need to update all of my sites
- Keep up with WordPress.
- Make it better than Drupal does (at the moment).

Re the last-mentioned point: automatic updates could be a game changer in Backdrop adoption from a Drupal 7 perspective. At the moment, I build only new sites with Backdrop but I don't consider to switch any sites from D7 to Backdrop because they're running fine. That would change definitely if we had automatic updates.

**Comment by quicksketch** [comment]
Some references:

- Here's the WordPress auto-updater code: https://github.com/WordPress/WordPress/blob/master/wp-admin/includes/class-wp-upgrader.php
- Amazingly comprehensive guide: https://paragonie.com/blog/2016/10/guide-automatic-security-updates-for-php-developers#elements-automatic-updates
  - This author has also attempted to help make automatic updates better for both WordPress (https://core.trac.wordpress.org/ticket/39309) and possible in Drupal (https://www.drupal.org/project/drupal/issues/2367319#comment-11415297)
  - Also, this gets a lot into digital signing and security of packages, which is a separate issue for us already at #1992, we might move the security and validation parts of this discussion over there.
- The Drush updater for Backdrop core: https://github.com/backdrop-contrib/drush/blob/1.x-0.x/commands/pm/backdrop_pm.drush.inc#L268

We may want to make a separate issue for "Let core update itself", as Update and Installer module combined can update and install updates for contrib modules already, but it cannot update core. Once we have core capable of updating itself, then we could build on top of that to make it happen automatically when new updates are available.

**Comment by serundeputy** [comment]
I've added https://github.com/backdrop/backdrop-issues/issues/3105 to allow Backdrop core updates via Backdrop UI.

**Comment by quicksketch** [comment]
I updated this issue to be a checkbox list of tasks.

**Comment by quicksketch** [comment]
New sub-issue: https://github.com/backdrop/backdrop-issues/issues/3208 Remove authorize.php entirely. I'm looking for more feedback on that proposal.

**Comment by jlfranklin** [comment]
I've added #3714 to the list.

**Comment by yorkshire-pudding** [comment]
I think we should remove the mention of automatic updates from the update settings as this is nowhere near "coming soon".
![image](https://user-images.githubusercontent.com/20357797/208666336-93dc5944-6daf-4809-bfd9-8ebb253849fc.png)

As @alanmels pointed out in #4793 this is confusing and potentially misleading, raising false expectations that are unlikely to be met in the short term.

**Comment by stpaultim** [comment]
We talked about this in meeting today. @quicksketch suggested that an audit of the Drupal module to see what might be useful for our purposes. This might be a relatively easy thing for someone to do and move this issue forward. 

If I understand correctly, we're mostly interested in finding a way to run update.php from the command line in core. @quicksketch can you add notes here about what needs to happen next?

**Comment by yorkshire-pudding** [comment]
Not automatically, but `bee` runs the update script from the command line.

https://github.com/backdrop-contrib/bee/blob/1.x-1.x/commands/update.bee.inc

**Comment by klonos** [comment]
> ...we're mostly interested in finding a way to run update.php from the command line in core.

I thought that the goal was to eventually allow a setting for automatic _**security**_ updates, so that people don't have to be waiting awake for announced security releases in order to patch their sites for 0-day vulnerabilities.

**Comment by yorkshire-pudding** [comment]
I think it would be good to have that as an option so you have:

1. Automatic Updates Off
2. Automatic Security Updates Only
3. Automatic Core Updates
4. Automatic Core and Contrib Updates

**Comment by stpaultim** [comment]
> I thought that the goal was to eventually allow a setting for automatic security updates, so that people don't have to be waiting awake for announced security releases in order to patch their sites for 0-day vulnerabilities.

Sorry, I was unclear and I have edited my comment. I was speaking about what we needed to do to move this issue forward, not the ultimate goal of the PR or the issue. 

I should have said, "If I understand correctly, to move forward we most need to find a way to run update.php from the command line in core."

**Comment by NormPlum** [comment]
I tend to agree with the views expressed in today's meeting about this issue: https://youtu.be/-jXdzcR4Ds4 Automatic updates aren't that important to me, and if they were available in core I'm not sure I would personally enable them...

I like that I get notified when there are updates. I can go and look over them to see what they are (what modules, big or small, etc.). I can check release notes if necessary, and then choose which modules to update manually, noting any steps I might need to take afterwards.

Nate asked us to consider not only our own personal situations, but those of the average site owner. I feel that the average site owner would not have the knowledge to fix issues as a result of modules being updated automatically - either bugs being introduced, or changes needing to be made to the site based on changes in how the module works. For those site owners who do have that knowledge, they can be updating manually anyway and the automatic nature would not be as much of a benefit to them I feel.

**Comment by olafgrabienski** [comment]
> I feel that the average site owner would not have the knowledge to fix issues as a result of modules being updated automatically ...

Agreed, as far as regular contrib updates are concerned. However, people could choose the option to perform automatic updates only in case of (critical) security issues for Backdrop core. If such an update were to break a site, the situation would still be better than with a hacked site.
